src/docs/arch-checkpoint.dox

```diff
@@ -0,0 +1,79 @@
+transaction. Each handle gathered refers to a specific b-tree. The set of b-trees gathered by the
+checkpoint varies based off configuration. Additionally clean b-trees, i.e. b-trees without any
+modifications are excluded from the list, with an exception for specific checkpoint configuration
+scenarios.
+
+2. The datafiles checkpoint
+The main work of checkpoint is done here, the array of b-tree's collected in the prepare stage are
+iterated over. For each b-tree, the tree is walked and all the dirty pages are reconciled. Skipping
+clean pages to avoid unnecessary work. Pages made clean ahead of the checkpoint by eviction are
+still skipped regardless of whether the update written by eviction is visible to the checkpoint
+transaction. The checkpoint guarantees that a clean version of every page in the tree exists and
+can be written to disk.
+
+3. The history store checkpoint
+The history store is checkpointed after the datafiles intentionally as during the reconciliation
+of the data files additional writes may be created in the history store and its important to include
+them in the checkpoint.
+
+4. Flushing the files to disk
+All the b-trees checkpointed and the history are flushed to disk at this stage, WiredTiger will wait
+until that process has completed to continue with the checkpoint.
+
+5. The metadata checkpoint
+A new entry into the metadata file is created for every datafile checkpointed, including the
+history store. As such the metadata file is the last file to be checkpointed. As WiredTiger
+maintains two checkpoints, the location of the most recent checkpoint is written to the turtle file.
+
+# Skipping checkpoints #
+
+It is possible that a checkpoint will be skipped. If no modifications to the database have been
+made since the last checkpoint, and the last checkpoint timestamp is equal to the current stable
+timestamp then a checkpoint will not be taken. This logic can be overridden by forcing a checkpoint
+via configuration.
+
+# Checkpoint generations #
+
+The checkpoint generation indicates which iteration of checkpoint a file has undergone, at the start
+of a checkpoint the generation is incremented. Then after processing any b-tree its
+\c checkpoint_gen to the latest checkpoint generation. Checkpoint generations impact visibility
+checks within WiredTiger, essentially if a b-tree is behind a checkpoint, i.e. its checkpoint
+generation is less than the current checkpoint generation, then the checkpoint transaction id and
+checkpoint timestamp are included in certain visibility checks. This prevents eviction from evicting
+updates from a given b-tree ahead of the checkpoint.
+
+# Garbage collection #
+
+While processing a b-tree checkpoint can mark pages as obsolete, any page that has an aggregated
+stop time pair which is globally visible will no longer be required by any reader and can be marked
+as deleted. This occurs prior to the page being reconciled, allowing for the page to be removed
+during the reconciliation.
```

Comment:
Can we add some more details for garbage collection like the removed page is not immediately available for reuse as it may be used in older checkpoints? Once the older checkpoint gets deleted it is added back to the free pages. Based on the free pages available at the file, the file compaction can happen?